src/reader.rs

```diff
@@ -0,0 +1,939 @@
+                    }
+                }
+            }
+        }
+
+        (found, count)
+    }
+
+    fn get_cycles_count(fun_edges: &mut [GcovEdge],
+                        fun_blocks: &[GcovBlock],
+                        blocks: &[usize]) -> u64 {
+        let mut count: u64 = 0;
+        let mut path: Vec<usize> = Vec::new();
+        let mut blocked: Vec<usize> = Vec::new();
+        let mut block_lists: Vec<Vec<usize>> = Vec::new();
+        for b in blocks {
+            path.clear();
+            blocked.clear();
+            block_lists.clear();
+            let (_, c) = GcovFunction::look_for_circuit(fun_edges, fun_blocks, *b, *b, &mut path, &mut blocked, &mut block_lists, blocks, count);
+            count += c;
+        }
+        count
+    }
+
+    fn get_line_count(fun_edges: &mut [GcovEdge],
+                      fun_blocks: &[GcovBlock],
+                      blocks: &[usize]) -> u64 {
+        let mut count: u64 = 0;
+        for b in blocks {
+            let block = &fun_blocks[*b];
+            if block.source.is_empty() {
+                count += block.counter;
+            } else {
+                for e in &block.source {
+                    let e = &fun_edges[*e];
+                    let w = e.source;
+                    if !blocks.iter().any(|x| *x == w) {
+                        count += e.counter;
+                    }
+                }
+            }
+            for e in &block.destination {
+                let e = &mut fun_edges[*e];
+                e.cycles = e.counter;
+            }
+        }
+
+        count + GcovFunction::get_cycles_count(fun_edges, fun_blocks, blocks)
+    }
+
+    fn add_line_count(&mut self) {
+        self.executed = self.edges.first().unwrap().counter > 0;
+        if self.executed {
+            let mut lines_to_block: HashMap<u32, Vec<usize>> = HashMap::new();
+            for block in &self.blocks {
+                let n = block.no;
+                for line in &block.lines {
+                    match lines_to_block.entry(*line) {
+                        hash_map::Entry::Occupied(vec) => {
+                            vec.into_mut().push(n);
+                        }
+                        hash_map::Entry::Vacant(v) => {
+                            v.insert(vec![n]);
+                        }
+                    }
+                }
+            }
+            let lines_to_block = lines_to_block;
+            for (line, blocks) in lines_to_block {
+                let count = GcovFunction::get_line_count(&mut self.edges, &self.blocks, &blocks);
+                self.lines.insert(line, count);
+            }
+        } else {
+            for block in &self.blocks {
+                for line in &block.lines {
+                    self.lines.entry(*line).or_insert(0);
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    
+    use super::*;
+    use std::path::PathBuf;
+
+    #[test]
+    fn test_reader_gcno() {
```

Comment:
Many parts of these tests are shared, maybe you can refactor them out in a function that you reuse in all tests?